# Saccharomyces cerevisiae bulk RNA seq analysis
A workflow outlining the steps to analyze bulk transcriptomic data on Saccharomyces cerevisiae

## Introduction
*Saccharomyces cerevisiae (S. cerevisiae)*, also known as brewer’s yeast, plays an important role in manufacturing alcoholic drinks such as beer and wine. In the context of sherry wine, flor yeast is a specialized type of *S. cerevisiae* that forms biofilms/velum on the surface of wine during the winemaking process. There are three biofilm stages: the early stage where it is characterized by the initial yeast attachment and adaptation, the thin stage where yeast cells are metabolically changing, and the mature stage define by a stable oxidative biofilm structure (Mardanov et al., 2020). Whereas most yeast strains are used for alcoholic fermentation, flor yeast is instead used to biologically age sherry wines. Therefore, flor yeast must adapt to use alternative modes of carbon and nitrogen metabolism to withstand the stressful environment where acetaldehyde and ethanol concentrations are high (Alexandre, 2013). In fact, ethanol and glycerol becomes the primary source of carbon. Characterizing these stage transitions with transcriptomics would provide insight into the molecular mechanisms that underlie flor yeast metabolism and survival. In fact, *S. cerevisiae* is a widely used and well-characterized model organism. It was the first eukaryotic model organism where RNA-sequencing has been employed (Nagalakshmi et al., 2008). Despite the advancements of RNA-sequencing technologies, analysis is required to fully understand the transcriptome of flor yeast. 

In fact, many computational tools have been developed to analyze RNA-seq datasets. Therefore, it is important to know the inherit advantages and disadvantages of each tool. Historically, trimming has been included as a preprocessing step in many standard RNA-seq pipelines. However, recent studies have reported that trimming is not necessarily required prior to gene counts due to many reasons. Firstly, a study reported that aggressive trimming with several trimming tools such as Trimmomatic (Bolger et al., 2014) in Drosophila melanogaster RNA-seq data reported significant variations of over ten percent in gene expression (Williams et al., 2016). In fact, over trimming may result in improper alignment of reads to the reference genome or transcriptome. In addition, it has been reported that untrimmed reads perform similarly or slightly better than trimmed reads (Liao & Shi, 2020). Therefore, trimming will not be utilized prior to processing gene counts. In fact, trimming may not be required if splice aware alignments are employed or if the quality of the reads is ideal.

If novel isoforms of genes transcripts are being studied, splice aware programs such as STAR (Spliced Transcripts Alignment to a Reference) (Dobin et al., 2013) and HISAT2 (hierarchical indexing for spliced alignment of transcripts 2) can be employed. In addition, it has been reported that splicing does occur in yeast however, its function is limited compared to higher order eukaryotic organisms and mainly results in nonsense-mediated mRNA decay (Kawashima et al., 2014). Minimap2 is also another option however, it has been suggested to be more suited to long-read sequencing data such as Oxford Nanopore or Pacific Biosciences Illumina (Patro et al., 2017). In contrast, if gene expression analysis is the main goal, pseudo-aligning programs such as Kallisto (Bray et al., 2016) or Salmon (Patro et al., 2017) can be used instead. The advantage of Salmon and Kallisto over STAR and HISAT2 is that they are computationally faster due to aligning to the transcriptome instead of aligning base to base to the reference genome. However, a disadvantage of alignment-free methods when compared to alignment-based methods is quantifying small RNAs such as tRNAs. In fact, there was a higher tRNA root mean square error (>0.4) for Kallisto and Salmon compared to alignment-based methods (<0.25) such as HISAT2 (Wu et al., 2018). When comparing both pseudo-alignment methods, Salmon has been reported to be more accurate than Kallisto due to the program’s inherit GC and sequencing bias detection which improves the accuracy and robustness of quantification (Patro et al., 2017). In fact, bias is an important aspect that needs to be considered when performing RNA-sequencing analysis since bias affects downstream analysis in the form of noise. Therefore, Salmon will be employed because of its accuracy, the main goal is to perform gene expression analysis in *S. cerevisiae*, and due to the limited splicing that occur in yeast.

Many packages are available in RStudio to perform downstream gene expression analysis and choosing which one to employ depends on the biological question, the type of data, and experimental design. For example, DESeq2 (Love et al., 2014) and edgeR (Robinson et al., 2010) are two commonly used programs to analyze RNA-seq data. Both tools use a negative binomial distribution model that generally perform well on 2 group pairwise comparison to maximize power. In fact, DESeq2 is commonly used for pairwise comparison analysis (Spies et al., 2019). Alternatively, limma is another RNA-seq package that uses linear models to analyze RNA-seq data (Ritchie et al., 2015). In fact, limma excels at time course-based experiments and more complex experimental designs. However, a study shows that short time course series which was defined as less than 8 timepoints, performed worse when time-course methods were used when compared to pairwise comparisons. In fact, DESeq2 which was used as the pairwise condition, exhibited 1014 true positives (TP), 70 false positives (FP), and a F1 score of 0.89 while most of the other time-course methods has worse summary statistics except for ImpulseDE2 (Spies et al., 2019). While both use robust normalization methods, with DESeq2 using median of ratios, and edgeR using trimmed mean of M-values, DESeq2 has been reported to perform slightly better than edgeR when sample sizes that are smaller than 10 (n < 10) due to its more conservative estimates resulting in lower false positives (Thawng & Smith, 2023). In fact, the median of ratios method assumes that not all genes are differentially expressed and is robust enough to detect if there are significant changes even when the data has much more upregulated genes than downregulated or vice-versa. Since this analysis will have a total of nine samples, with a total of 3 timepoints which is defined as a short time-course series, and were sampled randomly, DESeq2 will be employed to reduce the chance of FP discoveries and increase TP discoveries. While DESeq2 uses an adaptive Normally distributed prior to shrink the effect size of log fold changes (LFC), it has been reported that this method can over-shrink true positives of large LFCs or genes with adequate signal can be lost. Therefore, the Approximate Posterior Estimation for generalized linear model (apeglm) (Zhu et al., 2019), will be used. This method uses heavy tail priors while still preserving large effects, shrinkis noise from small estimates, and reduces high variance in LFC values.

The overall goal is to perform comparative transcriptomic analysis of a *S. cerevisiae* flor strain, I-329. Specifically, this analysis aims to characterize stage-specific transcriptional and functional pathway changes across early, thin, and mature biofilms to understand metabolic and stress-related pathways during sherry-like wine manufacturing. The analysis will be done by pairwise comparisons. Consequently, this allows for the counting and cataloguing of mRNA transcripts, and to characterize gene expression under different biofilm stages. In addition, this allows for a full and comprehensive transcriptomic analysis that includes differential gene expression (DEGs) and gene ontology (GO) analysis. 

## Methodology
#### Obtaining Transcriptome Data and Quality Control
RNA-sequencing data were downloaded from the European Nucleotide Archive (ENA) (https://www.ebi.ac.uk/ena/browser/home) using the accession numbers SRR10551657 - SRR10551665. The download files and Salmon analysis bash script code was adapted from the tutorial (https://combine-lab.github.io/salmon/getting_started/). Briefly, the samples were previously sequenced with the Illumina HiSeq 2500 platform and generated 6 to 9 million reads that were 50bp in length for each sample (Mardanov et al., 2020). The samples represent three different stages/groups of yeast biofilm formation: early biofilm, thin biofilm, and mature biofilm; with each group having a total of 3 replicates (Table 1). The *S. cerevisiae* strain S288C reference transcriptome file was downloaded from the NCBI’s database (https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000146045.2/). The S288C strain was used because its genes are well-annotated, well-maintained, and has the most complete assembly out of all strains which will allow for robust downstream gene expression analysis (Engel et al., 2014). Packages such as FastQC, MultiQC, and Salmon were downloaded via Bioconda (Grüning et al., 2018) on Bash terminal. Accurate and quick quality control is a major critical step to determine the quality of the RNA-sequencing reads. Therefore, FastQC v0.12.1 (Andrews, 2010) was employed to check the quality of the sequencing reads. Salmon v1.10.3 (Patro et al., 2017) was utilized to quantify the number of reads with the “-l A flag” for automatic detection of library type and the “--gcbias” flag as recommended by DESeq2’s vignette (Love et al., 2014). To check the quality of the quantification results from Salmon, MultiQC v1.33 (Ewels et al., 2016) was used to quickly assess important summary statistics such as percent of mapped reads and number of mapped reads. Quality control plots from MultiQC and FastQC were visualized using ggplot2 v4.0.2 (Wickham, 2009) in RStudio v4.5.1 (2025-06-13). 

#### Differential Gene Expression Analysis
Gene quantifications were imported intro RStudio. DESeq2 v1.50.2 (Love et al., 2014) package was employed for normalization by the medians of ratios method using the DEseq() function, and the default Wald test to obtain comparable gene expression levels between samples. Three pairwise comparisons were examined using the DESeqDataSet generated using the design formula ~stage to to compare early to thin, thin to mature, and early to mature stages. Therefore, apeglm v1.32.0 (Zhu et al., 2019) package was employed to increase the accuracy of LFC estimates that have low counts or high variability. Principal component analysis (PCA) was used to determine the data structure of each sample. The top 20 DEGs were filtered for adjusted p-value (padj) <0.05 and the absolute values of LFC of each gene in all three corresponding pairwise comparison were extracted. Independent scaling was done for each comparison. Key results include volcano plots for DEGs, PCA plot for data structure, and heatmap for the top 20 DEGs. 

#### GSEA and KEGG Analysis
Gene Set Enrichment Analysis (GSEA) was performed using annotations from the *S. cerevisiae* genome database (https://www.yeastgenome.org/) . Statistical analyses of GSEA were conducted with clusterProfiler v4.18.4 (Yu et al., 2012), with the application of Benjamini-Hochberg correction to control for false discovery rates (FDR), a padj cutoff of < 0.05, and fast gene set enrichment analysis (fgsea) (Korotkevich et al., 2019) method to identify enriched and significant pathways. GSEA analysis and plotting code were adapted from the tutorials available online (https://learn.gencore.bio.nyu.edu/rna-seq-analysis/gene-set-enrichment-analysis/) and (https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html). While Over-Representation Analysis (ORA) is a common method for gene analysis, GSEA is more powerful and robust because it considers all ranked distribution of genes rather than ORA’s method of using only top DEGs. GO annotations were retrieved from the genome wide yeast database using the org.Sc.gsd.db package within clusterProfiler, with a focus on biological process (BP) ontology. clusterProfiler’s simplify() function was applied to remove redundant GO terms. For GSEA, genes were ranked according to the log2foldchange values obtained from DESeq2 and sorted in descending order. To complement GO-based enrichment analysis, the Kyoto Encyclopedia of Genes and Genomes (KEGG) pathway enrichment was employed with the “enrichKEGG” function, with parameters organism = “sce” for *S.cerevisiae* to assess metabolic and signalling pathways. Key results were plotted as GSEA enrichment plots and KEGG dot plots.

All analysis scripts, including quality control, Salmon quantification, and R Studio code for gene expression and enrichment analysis is available at https://github.com/liona-vu/Saccharomyces_cerevisiae_bulk_RNA_seq_analysis. 

## Results
<img width="581" height="272" alt="Screenshot 2026-03-01 at 12 05 50 AM" src="https://github.com/user-attachments/assets/edf94b57-9a85-4cd2-a274-c3b977bbed9b" />

**Table 1. Metadata of the I-329 flor yeast samples.** Each accession number corresponds to a biofilm stage, with each stage containing 3 biological replicates (n=3) for a total of 9 samples. The information was adapted from the study by (Mardanov et al., 2020).

<img width="575" height="276" alt="Screenshot 2026-03-01 at 12 06 59 AM" src="https://github.com/user-attachments/assets/bee5bfa7-450f-450a-8028-6c223b4fd3c8" />

**Table 2. Summar statistics of SRR10551657 - SRR10551665 flor yeast samples.** Briefly, fastq sequencing files were quality control checked with FastQC. The table denotes the accession number, the total sequences, sequences flagged as poor quality, sequence length in bp, GC content, and the Phred score outputted across all 50 bp positions. 

<img width="2195" height="1384" alt="20260301_summary_statistics" src="https://github.com/user-attachments/assets/578b9ff4-16e6-44e0-94be-47226fa35e2f" />

**Figure 1. Important summary statistics outputted by MultiQC and FastQC.** Each barplots show the A. percent of mapped reads to the reference transcriptome in millions, and B. the number of mapped reads to the reference transcriptome. Plots were generated in RStudio using the ggplot package. Each bar represents different accession numbers.

#### Quality check of RNA-seq data reveals sufficient quality 
FastQC quality control revealed that RNA-seq data was of high quality. Specifically, no reads were flagged as poor quality, GC content ranged from 41% - 45% which is slightly higher than the average of 38.3% (Wood et al., 2002), and base sequence Phred scores were 30 (Table 2). Therefore, no trimming was required, and no samples were excluded from the analysis. Quality control summary statistics by MultiQC after Salmon quantification indicate all mapping rate percentages were greater than 80% for samples except for SRR10551660 and SRR10551661 which ranged around 74 – 76%, indicating a generally acceptable mapping rate Fig. 1A). A mapping rate of less than 50% would be concerning as it may indicate sample contamination, poor ribosomal RNA depletion, or problems with data processing (Dobin & Gingeras, 2015). In addition, Fig. 1B show the mapped reads which were in the millions for all samples. Overall, summary statistics show that the RNA-sequencing and mapping reads were ideal and acceptable for gene expression analysis.

**Figure 2. Volcano plots for DEGs in all three pairwise comparison of biofilm development.** Each plot represents A. early to thin, B. thin to mature, and C. early to mature. Red signifies DEGs that are upregulated, blue signifies DEGs that are downregulated, and grey signifies non-significant genes. DEGs were defined as significant if P-value is < 0.05, and absolute value of LFC is >1. 


**Figure 3. Heatmap of top 20 differentially expressed genes (DEGs) between all three pairwise comparisons of each biofilm developmental stage.** Each heatmap represents DEGs identified between A. early to thin, B. thin to mature, and C. early to mature stages. Rows represent genes that had the most significant LFC, had variance stabilizing transformation applied, and is scaled by row to highlight expression patterns. Colours on scales indicate that yellow represents higher expression and purple indicates lower expression. Columns represent individual biological replicates (n=3) and is marked by colour for each stage. Hierarchal clustering by biofilm stages on the columns is observed. 


<img width="1041" height="956" alt="20260301_Figure_04_PCA_plot" src="https://github.com/user-attachments/assets/18f9f2d5-ef9a-45f6-813e-e22d8c150add" />

**Figure 4. PCA plot of RNA-seq flor yeast data.** Principal component analysis (PCA) plot illustrating 9 flor yeast samples, showing clear distinctions for each developmental biofilm stages: early (red), mature (green), and thin (blue). Each point is labeled with its accession number. Principal component (PC) PC1 and PC2 explain 66% and 24% of variance, respectively. 

## Significant genes observed for all three stage comparisons
Volcano plots indicate that there are many significant genes that are either upregulated or downregulated across all three pairwise comparisons (Fig. 2A-C). However, this does not identify the names of the significant genes. Therefore, the top significant 20 genes were extracted after apeglm LFC shrinkage for each stage transition (Fig. 4A-C). It is important to note that since each heatmap is scaled independently, relative gene expression is only comparable within that specific comparison only and not throughout all 3 heatmaps. *FIT3* (Facilitator of Iron Transport 3) and PHR1 (PHotoreactivation Repair deficient 1) were upregulated in early samples relative to thin samples (Fig. 3A). In fact, FIT3 is a mannoprotein protein that helps with iron homeostasis and uptake in the yeast cell wall (Protchenko et al., 2001) while PHR1 responds to DNA damage (Sebastian & Sancar, 1991). *AHD7* (Alcohol DeHydrogenase 7), and *DAK2* (DihydroxyAcetone Kinase 2), genes that reduces and detoxifies aldehydes and detoxification of dihydroxyacetone, respectively, were highly expressed in thin samples compared to mature samples. 

**Figure 5. Gene set enrichment analysis (GSEA) enrichment plot for each biofilm comparisons groups.** Enrichment plots for A. Early to Thin, B. Thin to Mature, and C. Early to Mature. The curves show the enrichment score for the top 3 gene sets identified for each comparison. Each tick represents a gene that falls within the ranked list. The lower panel denotes the ranking metric of how each gene is ranked. 


**Figure 6. KEGG enrichment analysis for each stage comparison.** Dot plots for A. Early to Thin, B. Thin to Mature, and C. Early to Mature. The size of the dots represents the gene ratio associated with the KEGG term. The scale colours represent the adjusted P-value < 0.05 that has been corrected with Benjamini-Hochberg method. The size of the circle represents the gene ratio. The cluster show both upregulated and downregulated pathways.


